text_metrics/utils.py

```diff
@@ -271,7 +282,242 @@ def _join_surp(words: list[str], tok_surps: list[tuple[str, float]]):
+    """
+    model_variant = model_name.split("/")[-1]
+    if "gpt-neox" in model_variant:
+        tokenizer = GPTNeoXTokenizerFast.from_pretrained(model_name)
+    elif "gpt-neo" in model_variant or "gpt" in model_variant or "opt" in model_variant:
+        tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
+    elif "Llama" in model_variant:
+        assert (
+            hf_access_token is not None
+        ), f"Please provide the HuggingFace access token to load {model_name}"
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_name, use_fast=True, token=hf_access_token
+        )
+    elif "pythia" in model_variant:
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_name, revision=pythia_checkpoint, use_fast=True
+        )
+    elif "mamba" in model_variant:
+        tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
+    else:
+        raise ValueError("Unsupported LLM variant")
+
+    if "pythia" in model_variant:
+        model = GPTNeoXForCausalLM.from_pretrained(
+            model_name, revision=pythia_checkpoint
+        )
+    elif "mamba" in model_variant:
+        model = MambaForCausalLM.from_pretrained(model_name)
+    elif "Llama" in model_variant:
+        model = LlamaForCausalLM.from_pretrained(model_name, token=hf_access_token)
+    else:
+        model = AutoModelForCausalLM.from_pretrained(model_name)
+
+    model = model.to(device)
+
+    return tokenizer, model
+
+
+def surprise(
+    sentence: str,
+    model: Union[AutoModelForCausalLM, GPTNeoXForCausalLM],
+    tokenizer: Union[AutoTokenizer, GPTNeoXTokenizerFast],
+    model_name: str,
+    stride: int = 512,
+):
+    """This function calculates the surprisal of a sentence using a language model.
+    In case the sentence is too long, it is split into chunks while keeping previous context of STRIDE tokens.
+
+    Args:
+        sentence (str): The sentence for which the surprisal is calculated
+        model (Union[AutoModelForCausalLM, GPTNeoXForCausalLM]): The language model from which the surprisal is calculated
+        tokenizer (Union[AutoTokenizer, GPTNeoXTokenizerFast]): The tokenizer for the language model
+        model_name (str): The name of the language model (e.g "gpt2", "gpt-neo-125M", "opt-125m", "pythia-1b")
+        stride (int, optional): The number of tokens to keep as context. Defaults to 512.
+
+    Returns:
+        Tuple[np.ndarray, List[Tuple[int]]]: The surprisal values for each token in the sentence, the offset mapping
+        The offset mapping is a list of tuples, where each tuple contains the start and end character index of the token
+    """
+    model_variant = model_name.split("/")[-1]
+    with torch.no_grad():
+        all_log_probs = torch.tensor([], device=model.device)
+        offset_mapping = []
+        start_ind = 0
+        try:
+            max_ctx = model.config.max_position_embeddings
+        except:
+            max_ctx = 8192
```

Comment:
why this number?

Reply:
Arbitrary, mamba models don't have max_ctx. I'll make it way bigger 1e10 to make sure it's clear that it's not a real bound